fix: OnNetworkDespawn not called during shutdown

NoelStephensUnity opened the pull request:
When shutting down directly, NetworkObjects were not running through the OnNetworkDespawn process prior to being destroyed.  This fixes the issue by assuring NetworkObjects are despawned prior to being destroyed (if applicable).

[MTT-1542](https://jira.unity3d.com/browse/MTT-1542)

**Also Fixes:**
[MTT-1254](https://jira.unity3d.com/browse/MTT-1254)

_This was discovered by the samples group and relates to:_
[MTT-1472](https://jira.unity3d.com/browse/MTT-1472) 
<!-- When backporting is required please add the type:backport-release-<version> label to this PR. This should include all versions in which this should be backported to. -->

### PR Checklist
- [X] Have you added a backport label (if needed)? For example, the `type:backport-release-*` label. After you backport the PR, the label changes to `stat:backported-release-*`.
- [X] Have you updated the changelog? Each package has a `CHANGELOG.md` file.

## Changelog
### com.unity.netcode.gameobjects
- Fixed NetworkObjects not being despawned before they are destroyed during shutdown for client, host, and server instances.

## Testing and Documentation
* Includes integration test that confirms this fix applies to client, host, and server instances.

<!--  Uncomment and mark items off with a * if this PR deprecates any API:
### Deprecated API
- [ ] An `[Obsolete]` attribute was added along with a `(RemovedAfter yyyy-mm-dd)` entry.
- [ ] An [api updater] was added.
- [ ] Deprecation of the API is explained in the CHANGELOG.
- [ ] The users can understand why this API was removed and what they should use instead.
-->

NoelStephensUnity committed:
fix

Renaming and refactoring DestroyNonSceneObjects to be DespawnNetworkObjects as this is really what it is doing.  The destroy portion is either up to the handler, if an in-scene NetworkObject it lets the unloading of the scene destroy it, and the rest are always destroyed.

NoelStephensUnity committed:
fix

updates to the changes in NetworkSpawnManager

NoelStephensUnity committed:
fix

Adding a "ShutdownInProgress" flag to be able to easily check if we are shutting down.

Adding a check for the NetworkManager.LocalClient since we are despawning in the shutdown method.

Ignoring the de-parenting of NetworkObjects when shutting down (this will have to be discussed a bit).

NoelStephensUnity committed:
fix

Test fix:  I think this was just an unexposed bug in the NetworkRigidBody2D test, but will verify this before making this final.

Fixed some additional minor issues like checking for IsSpawned as opposed to NetworkManager.IsListening and IsSpawned as opposed to NetworkObject.IsSpawned.

NoelStephensUnity committed:
fix

Avoiding spam when despawning and shutting down.
Renaming DespawnNetworkObjects to DespawnAndDestroyNetworkObjects.
If a NetworkObject is not spawned it gets destroyed.

NoelStephensUnity committed:
test fix

Fixing issue with editor tests and the NetworkManager.Singleton instance.

LukeStampfli commented on com.unity.netcode.gameobjects/Components/NetworkRigidbody2D.cs:
```diff
@@ -32,7 +32,7 @@ private void Awake()
         private void FixedUpdate()
         {
-            if (NetworkManager.IsListening)
+            if (IsSpawned)
```
Is this change related to the fix addressed by this PR? I think we would also have to update NetworkRigidbody (not 2d) if this change is needed.

NoelStephensUnity replied:
Yes it is.  There are scenarios where NetworkManager can be null and fixed update will be invoked still for a few frames.  Either case, it is better to check the IsSpawned than NetworkManager.IsListening since NetworkManager.IsListening is the second to last thing to be set during shutdown.  Really, if you look over the Shutdown sequence you can see that by the time that is set everything else is shutdown and NetworkObjects have been despawned (depending upon whether it is an in-scene placed NetworkObject or a dynamically spawned/generated one determines how long the parent GameObject will persist).

NoelStephensUnity committed:
Merge branch 'develop' into fix/OnNetworkDespawn-Not-Called-During-Shutdown

NoelStephensUnity committed:
fix

Adjusting the logic a bit on getting the NetworkObject associated with a NetworkBehaviour.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -282,9 +282,13 @@ public NetworkObject NetworkObject
+                    (NetworkManager.Singleton != null && !NetworkManager.Singleton.ShutdownInProgress))
                 {
-                    NetworkLog.LogWarning($"Could not get {nameof(NetworkObject)} for the {nameof(NetworkBehaviour)}. Are you missing a {nameof(NetworkObject)} component?");
+                    if (NetworkLog.CurrentLogLevel < LogLevel.Normal)
```
Since this area has a high probability to generate a lot of spam, I reduced it to require the NetworkLog level be Developer as there can be several frames (during Update and FixedUpdate) where NetworkBehaviours will make checks that require a reference to the NetworkObject but it doesn't exist.

The question here is do we want to generate a warning here since NetworkObject can be null and should be valid to return a null value for m_NetworkObject.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -333,6 +333,9 @@ public IReadOnlyList<ulong> ConnectedClientsIds
         /// </summary>
         public bool IsConnectedClient { get; internal set; }
 
+
```
Really, we should have a few states for the NetworkManager to provide the NetworkManager's current state.
This is a minimal fix in order to determine if we are in the middle of a shutdown.

NoelStephensUnity committed:
revert and fix

After discussing this with Luke, it turned out there was a bug in the NetworkRigidBody2D where it was resetting the isKinematic to true.  This reverts the changes made to the NetworkRigidbody2DDynamicTest and fixes the bug with NetworkRigidBody2D.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -668,17 +676,21 @@ internal void OnDespawnObject(NetworkObject networkObject, bool destroyGameObjec
 
-            // Move child NetworkObjects to the root when parent NetworkObject is destroyed
-            foreach (var spawnedNetObj in SpawnedObjectsList)
+            // If we are shutting down the NetworkManager, then ignore resetting the parent
```
@MFatihMAR 
This was my solution to handle the issue where changing the parent would cause NetworkObject.OnTransformParentChanged to log an exception because NetworkManager is null during shutdown when the NetworkObjects are being despawned.  Adding an additional series of checks in NetworkObject.OnTransformParentChanged seemed less-than-optimal since this is specific to only despawning.

0xFA11 replied:
```cs
if (NetworkManager == null || !NetworkManager.IsListening)
{
    transform.parent = m_CachedParent;
    Debug.LogException(new NotListeningException($"{nameof(NetworkManager)} is not listening, start a server or host before reparenting"));
    return;
}
```

is this the check you're referring to?

if so, we could split that logic up into 2 pieces: `NetworkManager == null` and `NetworkManager.IsListening`.
we'd return early if `NetworkManager == null`
we'd log exception if `!NetworkManager.IsListening`

also, I think we should not introduce the `NetworkManager.ShutdownInProgress` (see my other comment above) and also not inject this extra check here.

NoelStephensUnity replied:
It is just below:
```
if (!NetworkManager.ShutdownInProgress)
{
```
Basically, if we a shutting down then ignore the parenting code...

NoelStephensUnity committed:
test

Tests that validate OnNetworkDespawn is invoked during shutdown.
This tests server, host, and client relative instances.

NoelStephensUnity committed:
update changelog

update the changelog with the fixed information

NoelStephensUnity committed:
Merge branch 'develop' into fix/OnNetworkDespawn-Not-Called-During-Shutdown

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -333,6 +333,9 @@ public IReadOnlyList<ulong> ConnectedClientsIds
         public bool IsConnectedClient { get; internal set; }
 
+
+        public bool ShutdownInProgress { get; internal set; }
```
Really, we should have a few states for the NetworkManager to provide the NetworkManager's current state.
This is a minimal fix in order to determine if we are in the middle of a shutdown.

0xFA11 replied:
instead of introducing this semi-fix/band-aid, what if we were to make `NetworkManager.Singleton = null` **after** we call `NetworkManager.Shutdown` (therefore destroy NOs first then set it to null)?

NoelStephensUnity replied:
That would work when there is a single instance running, but our MultiInstsance tests would fail.

0xFA11 replied:
huh? why?

NoelStephensUnity replied:
Because the NetworkManager.Singleton is one NetworkManager instance of (n) NetworkManager instances.
If you have 1 host and 4 clients running in a MultiInstance integration test and this code in NetworkManager is invoked during the shutdown of one of the clients:
```
        private void OnDestroy()
        {
            Shutdown();

            UnityEngine.SceneManagement.SceneManager.sceneUnloaded -= OnSceneUnloaded;

            if (Singleton == this)
            {
                Singleton = null;
            }
        }
```
But the NetworkManager of the client shutting down is not the same as the one assigned to the NetworkManager.Singleton, then it will not be set to null.  Forcing it to null in the integration/unit test could lead to other issues where code may depend upon the Singleton.  If only 1 client shuts down but the rest and the host stay active but we set the singleton to null... it would be not only introducing another semi-fix/band-aid...but the semi-fix/band-aid would be using a property for a purpose it was not intended and very likely would require us to fix other issues that might crop up by setting it to null while other NetworkManagers were still connected and active.

I didn't like adding yet another Boolean value to describe a "state of the NetworkManager". and is why I was thinking we might want to discuss coming up with some states for NetworkManager that let us know (some we already know but are various Boolean properties as opposed to just a single enum value):
- The NetworkManager is being started
- The NetworkManager is being initialized
- The NetworkManager is connected at the transport
- The NetworkManager is connected to network session
- The NetworkManager is disconnecting the transport
- The NetworkManager is shutting down
- The NetworkManager is shut down.
Perhaps not all of the above, but in that general direction.

0xFA11 replied:
I'm not suggesting to force `NetworkManager.Singleton = null`, I'm saying if we call network destroy after something sets `NetworkManager.Singleton = null` then let's call network destroy on objects first and later let that other code do `NetworkManager.Singleton = null`

NoelStephensUnity replied:
Well, either case it appears we have another [bool that signals the shutdown process that popped up recently](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/commit/5209fdca310fd7229f610d34bcdca3232f1965d2).
For now, I will just make it this: 
` public bool ShutdownInProgress { get { return m_ShuttingDown; } }`

I could make it internal but then users wouldn't be able to check this value for any edge case scenarios, which I could just use the m_ShuttingDown property but make it internal which would require me to change it to just ShuttingDown.

NoelStephensUnity committed:
Merge branch 'develop' into fix/OnNetworkDespawn-Not-Called-During-Shutdown

NoelStephensUnity committed:
Merge branch 'develop' into fix/OnNetworkDespawn-Not-Called-During-Shutdown

NoelStephensUnity committed:
Merge branch 'develop' into fix/OnNetworkDespawn-Not-Called-During-Shutdown

NoelStephensUnity committed:
Merge branch 'develop' into fix/OnNetworkDespawn-Not-Called-During-Shutdown

github-actions[bot] committed:
Merge develop into fix/OnNetworkDespawn-Not-Called-During-Shutdown

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/not_signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=1390) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=1390) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=1390) it.</sub>

andrews-unity committed:
Merge branch 'develop' into fix/OnNetworkDespawn-Not-Called-During-Shutdown

andrews-unity committed:
Merge branch 'develop' into fix/OnNetworkDespawn-Not-Called-During-Shutdown

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -282,9 +282,13 @@ public NetworkObject NetworkObject
                 }
 
-                if (m_NetworkObject == null && NetworkLog.CurrentLogLevel <= LogLevel.Normal)
+                if (m_NetworkObject == null || NetworkManager.Singleton == null ||
+                    (NetworkManager.Singleton != null && !NetworkManager.Singleton.ShutdownInProgress))
```
`NetworkManager.Singleton` check should be another if-check with potentially another log message IMO.

NoelStephensUnity replied:
Will revisit this.

NoelStephensUnity committed:
Merge branch 'develop' into fix/OnNetworkDespawn-Not-Called-During-Shutdown

NoelStephensUnity committed:
Merge branch 'develop' into fix/OnNetworkDespawn-Not-Called-During-Shutdown

NoelStephensUnity committed:
fix

fixing issue with test and recent changes for when the internal shutdown method is invoked.

NoelStephensUnity committed:
refactor

ShutdownInProgress now returns m_ShuttingDown.

Setting m_ShuttingDown to false at the end of ShutdownInternal in case the NetworkManager is going to be re-used.

NoelStephensUnity committed:
style

line endings

NoelStephensUnity committed:
Revert "style"

This reverts commit e2ca11a63bd7b10a88a319e0a838c072b0e6c578.

NoelStephensUnity committed:
fix

fixing timing related issue for NetworkObjectOnNetworkDespawnTests.
